### PR TITLE
[clang-doc] add ftime profiling

### PR DESCRIPTION
**Pull request by PeterChou1**
this patches adds ftime-tracing option for clang-doc which was what I used to profile the performance problems we had with clang-doc generating the llvm docs.
- ftime-trace  produces a time profiler json file called clang-doc-tracing.json

### CODE REVIEW

File: clang-tools-extra/clang-doc/tool/ClangDocMain.cpp [hunk 1]
```diff
@@ -316,32 +342,39 @@ Example usage for a project using a compile commands database:
...
         return;
       }
+      llvm::timeTraceProfilerEnd();
 
       // Add a reference to this Info in the Index
       {
         std::lock_guard<llvm::sys::Mutex> Guard(IndexMutex);
         clang::doc::Generator::addInfoToIndex(CDCtx.Idx, Reduced.get().get());
       }
-
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:331]: nit: unrelated change.

File: clang-tools-extra/clang-doc/tool/ClangDocMain.cpp [hunk 2]
```diff
@@ -362,6 +395,17 @@ Example usage for a project using a compile commands database:
...
+  llvm::timeTraceProfilerEnd();
+
+  if (FTimeTrace) {
+    std::error_code EC;
+    llvm::raw_fd_ostream OS("clang-doc-tracing.json", EC,
+                            llvm::sys::fs::OF_Text);
+    if (!EC) {
+      llvm::timeTraceProfilerWrite(OS);
+    } else {
+      llvm::errs() << "Error opening file: " << EC.message() << "\n";
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:407]: shouldn't we be returning an error value from main here?

File: clang-tools-extra/clang-doc/tool/ClangDocMain.cpp [hunk 3]
```diff
@@ -362,6 +395,17 @@ Example usage for a project using a compile commands database:
...
-
+  llvm::timeTraceProfilerEnd();
+
+  if (FTimeTrace) {
+    std::error_code EC;
+    llvm::raw_fd_ostream OS("clang-doc-tracing.json", EC,
+                            llvm::sys::fs::OF_Text);
+    if (!EC) {
+      llvm::timeTraceProfilerWrite(OS);
+    } else {
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:406]: nit: no braces for single statement conditions.

File: clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp [hunk 4]
```diff
@@ -30,7 +30,7 @@ ClangDocContext
 getClangDocContext(std::vector<std::string> UserStylesheets = {},
                    StringRef RepositoryUrl = "") {
   ClangDocContext CDCtx{
-      {}, "test-project", {}, {}, {}, RepositoryUrl, UserStylesheets};
+      {}, "test-project", {}, {}, {}, {}, {}, RepositoryUrl, UserStylesheets};
```
Comment (ilovepi) [inline clang-tools-extra/unittests/clang-doc/HTMLGeneratorTest.cpp:33]: If the new parameters go at the end, and have default values, then you don't need to updated this.

**Review by ilovepi — CHANGES_REQUESTED** [review]

File: clang-tools-extra/clang-doc/tool/ClangDocMain.cpp [hunk 5]
```diff
@@ -99,6 +100,16 @@ URL of repository that hosts code.
 Used for links to definition locations.)"),
                   llvm::cl::cat(ClangDocCategory));
 
+static llvm::cl::opt<bool> FTimeTrace("ftime-trace", llvm::cl::desc(R"(
+Turn on time profiler. Generates clang-doc-tracing.json)"),
+                                      llvm::cl::init(false),
+                                      llvm::cl::cat(ClangDocCategory));
+
+static llvm::cl::opt<int> FTimeGranularity("ftime-gran", llvm::cl::desc(R"(
```
[the remark this answers is not in the document]
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/tool/ClangDocMain.cpp]: I thought it be useful to set granularity for ftime trace, I've adjusted the pr to just set it as a constant

File: clang-tools-extra/clang-doc/BitcodeReader.cpp [hunk 6]
```diff
@@ -681,6 +683,7 @@ llvm::Error ClangDocBitcodeReader::readRecord(unsigned ID, Reference *I) {
 // Read a block of records into a single info.
 template <typename T>
 llvm::Error ClangDocBitcodeReader::readBlock(unsigned ID, T I) {
+  llvm::TimeTraceScope("readBlock", "ClangDocBitcodeReader");
```
[the remark this answers is not in the document]
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/BitcodeReader.cpp]: there's really no logic I've updated the pr, so that it follow the convention of naming it after the scope or method the timescope is in

File: clang-tools-extra/clang-doc/Mapper.cpp [hunk 7]
```diff
@@ -30,6 +35,7 @@ template <typename T> bool MapASTVisitor::mapDecl(const T *D) {
   if (D->getParentFunctionOrMethod())
     return true;
 
+  llvm::timeTraceProfilerBegin("emit info phase", "emit info");
```
[the remark this answers is not in the document]
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/Mapper.cpp]: emit info was originally meant to mean the code path where the mapper serialized the ASTNode into clang-docs Info record

File: clang-tools-extra/clang-doc/Mapper.cpp [hunk 8]
```diff
@@ -13,12 +13,17 @@
...
 #include "llvm/ADT/StringExtras.h"
 #include "llvm/Support/Error.h"
+#include "llvm/Support/TimeProfiler.h"
 
 namespace clang {
 namespace doc {
 
 void MapASTVisitor::HandleTranslationUnit(ASTContext &Context) {
+  if (CDCtx.FTimeTrace)
+    llvm::timeTraceProfilerInitialize(CDCtx.Granularity, "clang-doc");
```
[the remark this answers is not in the document]
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/Mapper.cpp]: I thought it would be more clear since we also use llvm::timeTraceProfilerFinishThread in the method

[the remark this answers is not in the document]
Comment (ilovepi) [reply clang-tools-extra/clang-doc/Mapper.cpp]: There’s lots of ways to initialize/teardown things so I’m fine with it the way it is, but I wanted to know if there was a reason it wasn’t in those other places. 

I’d take another look at how this gets used in other tools, and follow whatever seems to be common.

File: clang-tools-extra/clang-doc/BitcodeReader.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ilovepi) [reply clang-tools-extra/clang-doc/BitcodeReader.cpp]: Have you looked at the profile with the new names? Is it easy to follow? I’d expect it’s a bit harder now, since you’ve dropped some info. 

From what I can see, the typical usage of the name is something like “parse input files”. Mostly within our tools we seem to try to capture how long some task is taking, e.g. parsing files, constructing the ast, generating call graphs, writing out files, etc. 

I’d consider structuring these more around the tasks your tracking, and perhaps use the details to track more specific bits within a larger task or phase.

[the remark this answers is not in the document]
Comment (PeterChou1) [reply clang-tools-extra/clang-doc/BitcodeReader.cpp]: I renamed most the labels to follow the convention of [task we are tracking] -> [details]

File: clang-tools-extra/clang-doc/BitcodeReader.cpp [hunk 9]
```diff
@@ -670,6 +671,7 @@ llvm::Error ClangDocBitcodeReader::readRecord(unsigned ID, T I) {
 
 template <>
 llvm::Error ClangDocBitcodeReader::readRecord(unsigned ID, Reference *I) {
+  llvm::TimeTraceScope("Reducing infos", "readRecord");
```
Comment (ilovepi) [inline clang-tools-extra/clang-doc/BitcodeReader.cpp:674]: I know this is terminology already in clang-doc, but lets come up w/ something a bit more informative than "Reducing Infos". TBH, the whole "Infos" bit has always rubbed me the wrong way, so even just `Reducing Results` is better IMO.  Use your judgement, but I think we can find a better name for these than `Infos`.

**Review by ilovepi — APPROVED** [review]
LGTM modulo a small nit in the naming, we can always add more instrumentation if we need it, but this is a good start.
